### PR TITLE
ROX-15348: Fix flaky Splunk integration test

### PR DESCRIPTION
**Pull request by Maddosaurus**
### Description
The `IntegrationsSplunkViolationsTest` has been flaky. 
If we set aside situations where the image pull did not work, most of the flakes can be attributed to missing alerts. 
Splunk Alerts are an integral part of our Splunk TA, therefore we have been asserting on them.
However, we did that in a brittle way.
This PR introduces a more robust way to assert that Splunk generates the required alerts.
 
#### Background  & Dataflow
The dataflow is as follows: 
Splunk queries ACS for all Violations it knows in regular intervals.
The Violations arrive in Splunk as regular events and are being indexed (query `search sourcetype=stackrox-violations`). 
Every 5 minutes, a Splunk Cronjob runs that converts all indexed Violations into Alerts.
Alerts from any source in Splunk can be queried via a specialized query (query `| from datamodel Alerts.Alerts`).

#### Introduced change
Before, we installed the Splunk TA for RHACS and configured it. After generating Violations in ACS, we directly asserted on Alerts, expecting that they exist in Splunk:
`Set up TA -> Generate Violations in ACS -> Assert on Splunk Alerts`

This PR breaks the test up into smaller steps:
`Set up TA -> Generate Violations in ACS -> Assert on Splunk indexed violations -> Trigger conversion cronjob -> Assert on Splunk Alerts`.

### User-facing documentation
- [x] CHANGELOG update is not needed
- [x] Documentation is not needed

### Testing
- [x] inspected CI results

#### Automated testing
- [x] modified existing tests

#### How I validated my change
Repeated CI tests, plus manual runs of the groovy test against GKE and OpenShift infra clusters

### CODE REVIEW

**Comment by Maddosaurus** [comment]
@rhacs-bot /help

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 1]
```diff
@@ -134,19 +132,64 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
             }
             hasNetworkViolation = results.any { isNetworkViolation(it) }
             hasProcessViolation = results.any { isProcessViolation(it) }
+            // TODO: Remove debug log
+            log.info "Found violations in Splunk: \n${results}"
+            log.info "hasNetworkViolation: ${hasNetworkViolation}\nhasProcessViolation: ${hasProcessViolation}"
+            // TODO: /Remove debug log
```
Comment (vikin91) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:138]: TODOs left - maybe they should stay?

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 2]
```diff
@@ -134,19 +132,64 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
             if (hasNetworkViolation && hasProcessViolation) {
                 log.info "Success!"
                 break
             }
         }
 
+        // After we know that violations arrived, manually kick off search that converts them into alerts
```
Comment (vikin91) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:145]: Let's turn that into info log with a milestone message - violations arriving looks like an important milestone to reach

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 3]
```diff
@@ -134,19 +132,64 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
+        postToSplunk(port, "/services/saved/searches/" +
+                "Threat%20-%20Create%20Notable%20from%20RHACS%20Alert%20-%20Rule/dispatch", [
```
Comment (vikin91) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:147]: Would that be a good candidate for a const with some description what that URL actually is?

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 4]
```diff
@@ -134,19 +132,64 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
+                "force_dispatch": "true",
+        ])
+
+        // Check for Alerts
+        List<Map<String, String>> alerts = Collections.emptyList()
+        boolean hasNetworkAlert = false
+        boolean hasProcessAlert = false
+        for (int i = 0; i < 20; i++) {
+            log.info "Attempt ${i} to get Alerts from Splunk"
+            // Hint: If this produces no results, evaluate expanding the earliest_time, e.g. to -15m
```
Comment (vikin91) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:158]: and where could I find the `earliest_time`?

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 5]
```diff
@@ -134,19 +132,64 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
+            alerts = vResponse.getBody().jsonPath().getList("results")
+            if (alerts.isEmpty()) {
+                continue
+            }
+            hasNetworkAlert = alerts.any { isNetworkViolation(it) }
+            hasProcessAlert = alerts.any { isProcessViolation(it) }
+            // TODO: Remove debug log
+            log.info "Found Alerts in Splunk: \n${alerts}"
+            log.info "hasNetworkAlert: ${hasNetworkAlert}\nhasProcessAlert: ${hasProcessAlert}"
+            // TODO: /Remove debug log
```
Comment (vikin91) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:175]: Same TODO - this could be easily `log.debug` and could potentially stay. WDYT?

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 6]
```diff
@@ -134,19 +132,64 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
+            if (hasNetworkAlert && hasProcessAlert) {
+                log.info "Success!"
```
Comment (vikin91) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:177]: Could we write a bit more what kind of success is this? Reading the code gives some hints, but let's make sure that reading the logs give the same - for example: "Success: Found network and process alert".

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 7]
```diff
@@ -134,19 +134,60 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
+        ])
+
+        // Check for Alerts
+        List<Map<String, String>> alerts = Collections.emptyList()
+        boolean hasNetworkAlert = false
+        boolean hasProcessAlert = false
+        for (int i = 0; i < 20; i++) {
+            log.info "Attempt ${i} to get Alerts from Splunk"
+            // Hint: If this produces no results, evaluate expanding the earliest_time, e.g. to -15m.
+            // This can be done by expanding `createSearch` with a new parameter.
```
Comment (jschnath) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:158]: Would this be something that would make sense for the test itself to do? Eg. retry with -15m on earliest_time if no results are produced? Under what circumstances would this happen, is it a rare edge case?

Comment (Maddosaurus) [reply qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:158]: I consider this the next course of action should the test still be flaky after these changes.
Ideally, we would never need to implement this. The test retries the current query over multiple minutes, so it should already catch all violations should there be any.

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 8]
```diff
@@ -119,9 +119,9 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
         boolean hasNetworkViolation = false
         boolean hasProcessViolation = false
         def port = splunkDeployment.splunkPortForward.getLocalPort()
-        for (int i = 0; i < 15; i++) {
-            log.info "Attempt ${i} to get violations from Splunk"
-            def searchId = SplunkUtil.createSearch(port, "| from datamodel Alerts.Alerts")
+        for (int i = 0; i < 20; i++) {
```
Comment (lvalerom) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:122]: I think you can use the `withRetry` helper function here instead of a for loop.

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 9]
```diff
@@ -134,19 +134,60 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
+                log.info "Success finding Network and Process Violations in Splunk!"
+                break
+            }
+        }
+
+        log.info "Starting conversion of ACS violations to Splunk alerts"
+        postToSplunk(port, "/services/saved/searches/" + SPLUNK_TA_CONVERSION_JOB_NAME +"/dispatch", [
+                "dispatch.now": "true",
+                "force_dispatch": "true",
+        ])
```
Comment (lvalerom) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:149]: If I'm not mistaken this will trigger the conversion, right? Does that mean that the logic inside the splunk TA that will trigger this conversion automatically is not tested?

I wonder if just setting the interval to `1m` (for example) and waiting in the next block would be sufficient without having to manually trigger the conversion.

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 10]
```diff
@@ -134,19 +134,60 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
+
+        // Check for Alerts
+        List<Map<String, String>> alerts = Collections.emptyList()
+        boolean hasNetworkAlert = false
+        boolean hasProcessAlert = false
+        for (int i = 0; i < 20; i++) {
+            log.info "Attempt ${i} to get Alerts from Splunk"
```
Comment (lvalerom) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:156]: Same as above. I think you can use `withRetry`

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 11]
```diff
@@ -134,19 +134,60 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
             }
             hasNetworkViolation = results.any { isNetworkViolation(it) }
             hasProcessViolation = results.any { isProcessViolation(it) }
```
Comment (lvalerom) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:136]: How do we know for sure that there will be network or process violations? Does this test create deployments that trigger these kind of violations?

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy (hunk 9 above)
Comment (Maddosaurus) [reply qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:149]: Correct, this triggers the conversion. 
I originally set this to 1m [in this commit](https://github.com/stackrox/stackrox/pull/11848/commits/405c477de9a304af25478ae2f87b81593ad0c7e2#diff-b5d8e8c78f856dbec06efbf9fdf975540853d6691648c5071a853559e9e47a5bL103), but after discussions in last weeks demo meeting, I reverted that change. 
It was argued that changing the cron would bring this further away from prod use.

I am triggering the same conversion job that is the cronjob. I would argue that we shouldn't do functional testing for the Splunk cronjob. We could introduce a third loop in the middle that just waits for the cronjob to have run, although I don't know whether we really should blow up the test runtime any further

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy (hunk 11 above)
Comment (Maddosaurus) [reply qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:136]: Yes, see [lines 346, ff](https://github.com/stackrox/stackrox/pull/11848/files#diff-b5d8e8c78f856dbec06efbf9fdf975540853d6691648c5071a853559e9e47a5bR346-R352). The test creates deployments that result in process and network violations.

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy (hunk 8 above)
Comment (Maddosaurus) [reply qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:122]: Using `withRetry` here would loose me the early continue in case we get no results back, correct?
Given that this is happening frequently (usually we get results after 4-5 tries), I lean towards keeping both loops. 
WDYT?

Comment (lvalerom) [reply qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:122]: 🤔 I think you can achieve a similar behavior just something like:
```groovy
withRetry(20, 15) {
    response = getSearchResults()
    assert response != null
    // Check for violations
    assert violations
}
```
If the response is `null` we early continue by failing the assertion.

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy (hunk 9 above)
Comment (lvalerom) [reply qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:149]: I don't have enough knowledge on the Splunk TA but they way I see it is:

- If the cronjob scheduler is something that splunk offers (not our code) and simply calls our conversion function, then I think it's ok to trigger the conversion manually.
- If the cronjob is our code, then I think it would be nice to verify that it's working by waiting until it triggers the conversion.

**Comment by Maddosaurus** [comment]
The failed e2e run for 5b423cc is something unrelated failing, not the Splunk test (NetworkFlowTest

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy (hunk 9 above)
Comment (Maddosaurus) [reply qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:149]: It is the first one. Splunk offers its own cron scheduler that calls a custom conversion function that we have built :smile:

**Comment by Maddosaurus** [comment]
Run f9e01e6 also did not fail the Splunk integration test, but the ImageScanningTest.

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 12]
```diff
@@ -119,34 +120,61 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
         then:
         "StackRox violations are in Splunk"
         assert !results.isEmpty()
+        assert !alerts.isEmpty()
         assert hasNetworkViolation
         assert hasProcessViolation
-        for (result in results) {
-            validateCimMappings(result)
+        assert hasNetworkAlert
+        assert hasProcessAlert
```
Comment (lvalerom) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:174]: If we got to this point I think these assertion are redundant, right?

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy [hunk 13]
```diff
@@ -119,34 +120,61 @@ class IntegrationsSplunkViolationsTest extends BaseSpecification {
...
-                log.info "Success!"
-                break
-            }
+            log.debug "Found violations in Splunk: \n${results}"
+            log.info "Current Splunk index contains " +
+                    "any Network Violation: ${hasNetworkViolation} and any Process Violation: ${hasProcessViolation}"
+            assert hasNetworkViolation && hasProcessViolation
+        }
+
+        log.info "Starting conversion of ACS violations to Splunk alerts"
```
Comment (lvalerom) [inline qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:139]: [nit] Super nitpicking but maybe add a comment or change the log to reflect that we are forcing the conversion that otherwise will be done by the cronjob.

**Review by lvalerom — APPROVED** [review]
I've added a few optional things but other than that looks good to me.

File: qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy (hunk 12 above)
Comment (lvalerom) [reply qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:174]: ☝️ the diff looks weird in this view. I meant these assertions:

```groovy
        assert !results.isEmpty()
        assert !alerts.isEmpty()
        assert hasNetworkViolation
        assert hasProcessViolation
        assert hasNetworkAlert
        assert hasProcessAlert
```

Comment (Maddosaurus) [reply qa-tests-backend/src/test/groovy/IntegrationsSplunkViolationsTest.groovy:174]: Good point. I will leave in the assertion for alerts not being empty, as we're passing that to `validateCimMapings`, but everything else has been asserted true already. So:
```groovy
        assert !alerts.isEmpty()
	log.info "Validating CIM mappings for alerts"
        for (alert in alerts) {
            validateCimMappings(alert)
	}
```
